File: android/app/src/main/java/com/bluelink/android/domain/SearchExcerpt.kt
package com.bluelink.android.domain

/** Display-only slices; actions must continue to use the original message or filename. */
data class SearchExcerpt(val text: String, val highlights: List<IntRange>) {
    companion object {
        fun create(source: String, query: String, budget: Int = 96, fileName: Boolean = false): SearchExcerpt {
            val bounds = Regex("\\X").findAll(source).map { it.range.first }.toList() + source.length
            val count = bounds.size - 1
            val limit = budget.coerceAtLeast(1)
            val needle = query.trim()
            val match = if (needle.isEmpty()) -1 else source.indexOf(needle, ignoreCase = true)
            val unit = if (match < 0) 0 else bounds.indexOfLast { it <= match }.coerceAtMost(count)
            val dot = if (fileName) source.lastIndexOf('.').takeIf { it > 0 && it < source.lastIndex } else null
            val suffix = dot?.let { bounds.indexOfLast { b -> b <= it } } ?: count
            val suffixUnits = count - suffix
            val reserve = if (count > limit && suffixUnits in 1 until limit / 2) suffixUnits else 0
            val window = (limit - reserve).coerceAtLeast(1)
            val start = if (count <= limit) 0 else (unit - minOf(12, window / 4)).coerceAtLeast(0)
            val end = if (count <= limit) count else (start + window).coerceAtMost(count)
            val parts = mutableListOf(bounds[start] to bounds[end])
            if (reserve > 0 && suffix >= end && end < count) parts += bounds[suffix] to source.length
            val matches = mutableListOf<IntRange>()
            if (needle.isNotEmpty()) {
                var index = source.indexOf(needle, ignoreCase = true)
                while (index >= 0) {
                    matches += index until index + needle.length
                    index = source.indexOf(needle, index + needle.length, ignoreCase = true)
                }
            }
            val output = StringBuilder()
            val spans = mutableListOf<IntRange>()
            var previous = 0
            for ((from, to) in parts) {
                if (from > previous) output.append('…')
                val offset = output.length
                output.append(source.substring(from, to).replace('\r', ' ').replace('\n', ' '))
                for (hit in matches) {
                    val left = maxOf(from, hit.first)
                    val right = minOf(to, hit.last + 1)
                    if (left < right) spans += (offset + left - from) until (offset + right - from)
                }
                previous = to
            }
            if (previous < source.length) output.append('…')
            return SearchExcerpt(output.toString(), spans)
        }
    }
}

File: windows/BlueLink.App/Domain/SearchExcerpt.cs
using System.Globalization;
using System.Text;

namespace BlueLink.Domain;

internal sealed record SearchExcerpt(string Text, IReadOnlyList<(int Start, int Length)> Highlights)
{
    internal static SearchExcerpt Create(string source, string query, int budget = 96, bool fileName = false)
    {
        var bounds = StringInfo.ParseCombiningCharacters(source).Append(source.Length).ToArray();
        var count = bounds.Length - 1;
        var limit = Math.Max(1, budget);
        var needle = query.Trim();
        var match = needle.Length == 0 ? -1 : source.IndexOf(needle, StringComparison.OrdinalIgnoreCase);
        var unit = match < 0 ? 0 : Math.Min(count, Array.FindLastIndex(bounds, b => b <= match));
        var dot = fileName ? source.LastIndexOf('.') : -1;
        var suffix = dot > 0 && dot < source.Length - 1 ? Array.FindLastIndex(bounds, b => b <= dot) : count;
        var suffixUnits = count - suffix;
        var reserve = count > limit && suffixUnits > 0 && suffixUnits < limit / 2 ? suffixUnits : 0;
        var window = Math.Max(1, limit - reserve);
        var start = count <= limit ? 0 : Math.Max(0, unit - Math.Min(12, window / 4));
        var end = count <= limit ? count : Math.Min(count, start + window);
        var parts = new List<(int From, int To)> { (bounds[start], bounds[end]) };
        if (reserve > 0 && suffix >= end && end < count) parts.Add((bounds[suffix], source.Length));
        var matches = new List<(int Start, int End)>();
        if (needle.Length > 0)
            for (var index = source.IndexOf(needle, StringComparison.OrdinalIgnoreCase); index >= 0;
                 index = source.IndexOf(needle, index + needle.Length, StringComparison.OrdinalIgnoreCase))
                matches.Add((index, index + needle.Length));
        var text = new StringBuilder();
        var spans = new List<(int Start, int Length)>();
        var previous = 0;
        foreach (var (from, to) in parts)
        {
            if (from > previous) text.Append('…');
            var offset = text.Length;
            text.Append(source[from..to].Replace('\r', ' ').Replace('\n', ' '));
            foreach (var hit in matches)
            {
                var left = Math.Max(from, hit.Start);
                var right = Math.Min(to, hit.End);
                if (left < right) spans.Add((offset + left - from, right - left));
            }
            previous = to;
        }
        if (previous < source.Length) text.Append('…');
        return new(text.ToString(), spans);
    }
}
